x-pack/filebeat/input/gcs/scheduler.go

```diff
@@ -0,0 +1,229 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package gcs
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+
+	cursor "github.com/elastic/beats/v7/filebeat/input/v2/input-cursor"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-concert/timed"
+)
+
+// limiter, is used to limit the number of goroutines from blowing up the stack
+type limiter struct {
+	ctx context.Context
+	wg  sync.WaitGroup
+	// limit specifies the maximum number
+	// of concurrent jobs to perform.
+	limit chan struct{}
+}
+type scheduler struct {
+	publisher cursor.Publisher
+	bucket    *storage.BucketHandle
+	src       *Source
+	cfg       *config
+	state     *state
+	log       *logp.Logger
+	limiter   *limiter
+}
+
+// newScheduler, returns a new scheduler instance
+func newScheduler(publisher cursor.Publisher, bucket *storage.BucketHandle, src *Source, cfg *config,
+	state *state, log *logp.Logger,
+) *scheduler {
+	return &scheduler{
+		publisher: publisher,
+		bucket:    bucket,
+		src:       src,
+		cfg:       cfg,
+		state:     state,
+		log:       log,
+		limiter:   &limiter{limit: make(chan struct{}, src.MaxWorkers)},
+	}
+}
+
+// Schedule, is responsible for fetching & scheduling jobs using the workerpool model
+func (s *scheduler) schedule(ctx context.Context) error {
+	s.limiter.ctx = ctx
```

Comment:
Why is this split out here?

Reply:
fixed the structure accordingly. The job needs the parent ctx and not the one with the timeout so , the timeout ctx ius required only by gcs sdk calls.